patroni/quorum.py

```diff
@@ -0,0 +1,427 @@
+
+        :raises:
+            :exc:`QuorumError` in case of invalid data or if the invariant after transition could not be satisfied.
+        """
+        if quorum < 0:
+            raise QuorumError(f'Quorum {quorum} < 0 of ({voters})')
+        if quorum > 0 and quorum >= len(voters):
+            raise QuorumError(f'Quorum {quorum} >= N of ({voters})')
+
+        old_leader = self.leader
+        if leader is not None:  # Change of leader was requested
+            self.leader = leader
+        elif self.numsync_confirmed == 0:
+            # If there are no nodes that known to caught up with the primary we want to reset quorum/votes in /sync key
+            quorum = 0
+            voters = CaseInsensitiveSet()
+        elif adjust_quorum:
+            # It could be that the number of nodes that are known to catch up with the primary is below desired numsync.
+            # We want to increase quorum to guaranty that the sync node will be found during the leader race.
+            quorum += max(self.numsync - self.numsync_confirmed, 0)
+
+        if (self.leader, quorum, voters) == (old_leader, self.quorum, self.voters):
+            if self.voters:
+                return
+            # If transition produces no change of leader/quorum/voters we want to give a hint to
+            # the caller to fetch the new state from the database and restart QuorumStateResolver.
+            yield Transition('restart', self.leader, self.quorum, self.voters)
+
+        self.quorum = quorum
+        self.voters = voters
+        self.check_invariants()
+        logger.debug('quorum %s %s %s', self.leader, self.quorum, self.voters)
+        yield Transition('quorum', self.leader, self.quorum, self.voters)
+
+    def sync_update(self, numsync: int, sync: CaseInsensitiveSet) -> Iterator[Transition]:
+        """Updates :attr:`numsync` and :attr:`sync` fields.
+
+        :param numsync: the new value for :attr:`numsync`.
+        :param sync: the new value for :attr:`sync`:
+
+        :yields: the new state of ``synchronous_standby_names`` as a :class:`Transition` object.
+
+        :raises:
+            :exc:`QuorumError` in case of invalid data or if invariant after transition could not be satisfied
+        """
+        if numsync < 0:
+            raise QuorumError(f'Sync {numsync} < 0 of ({sync})')
+        if numsync > len(sync):
+            raise QuorumError(f'Sync {numsync} > N of ({sync})')
+
+        self.numsync = numsync
+        self.sync = sync
+        self.check_invariants()
+        logger.debug('sync %s %s %s', self.leader, self.numsync, self.sync)
+        yield Transition('sync', self.leader, self.numsync, self.sync)
+
+    def __iter__(self) -> Iterator[Transition]:
+        """Merge two transitions of the same type to a single one.
+
+        .. note::
+            This is always safe because skipping the first transition is equivalent
+            to no one observing the intermediate state.
```

Comment:
```suggestion
        """Iterate over the transitions produced by :meth:`_generate_transitions`.

        .. note::
            Merge two transitions of the same type to a single one.

            This is always safe because skipping the first transition is equivalent
            to no one observing the intermediate state.
```

This line seems more like a note about the method than a description of what it is intended to do.

What about this?